### PR TITLE
 Target name should not be the same as an executable product name (by case-insensitive comparison)

### PR DESCRIPTION
**Pull request by elsh**
Target names differing from an executable name only in casing can cause a lot of issues. In such case, they are now considered duplicates (as of toolchain >= v5.5) and manifest validation will throw an error. 
Fixes rdar://77466469.

### CODE REVIEW

**Comment by elsh** [comment]
@swift-ci smoke test

File: Sources/PackageLoading/ManifestLoader.swift [hunk 1]
```diff
@@ -331,6 +333,12 @@ public final class ManifestLoader: ManifestLoaderProtocol {
 
             // Check that the product references existing targets.
             for target in product.targets {
+                // Check if this target has the same name as an executable product name (by case-insensitive comparison)
+                if manifest.toolsVersion >= .v5_5,
+                   execNames.contains(target.lowercased()) {
```
Comment (elsh) [inline Sources/PackageLoading/ManifestLoader.swift:338]: Are there issues like this for non-executable products as well?  If so, I can update this to check against all products.

File: Sources/PackageLoading/ManifestLoader.swift [hunk 2]
```diff
@@ -322,6 +322,8 @@ public final class ManifestLoader: ManifestLoaderProtocol {
     }
 
     private func validateProducts(_ manifest: Manifest, diagnostics: DiagnosticsEngine?) throws {
+        let execNames = manifest.products.filter { $0.type == .executable }.map { $0.name }.map { $0.lowercased() }
```
Comment (tomerd) [inline Sources/PackageLoading/ManifestLoader.swift:325]: combine two maps into one?

Comment (tomerd) [reply Sources/PackageLoading/ManifestLoader.swift:325]: also maybe use a set for faster lookup?

File: Sources/PackageLoading/ManifestLoader.swift (hunk 1 above)
Comment (tomerd) [reply Sources/PackageLoading/ManifestLoader.swift:338]: @neonichu I am also a little confused by the exact requirements here. isn't a target and product having the same name pretty typical? in many cases they are 1:1 and have the same name

Comment (abertelrud) [reply Sources/PackageLoading/ManifestLoader.swift:338]: It's quite typical.  I don't think that this change is correct; I think that the conclusion from the bug report was that SwiftPM should guard against two targets have the same case-insensitive name, not whether a product and a target have the same name.

Comment (neonichu) [reply Sources/PackageLoading/ManifestLoader.swift:338]: Yep, that's right. We probably also should disallow having two products having the same name just differing by case, but a target and a product having the same name is OK (and even the default from `init` IIRC).

Comment (elsh) [reply Sources/PackageLoading/ManifestLoader.swift:338]: Targets in the bug report don't have the same case-insensitive name (see the `testTargetNameAcrossProductDecls`). I'll look into an alternative way to resolve this, e.g. no --module-name for executables.

**Comment by elsh** [comment]
Closing this as the build results seem to be correct on the SwiftPM side; there's a discrepancy with xcodebuild to be addressed separately.

**Pull request closed without merging**